### ISSUE TITLE
Configurable lint rule for unsafe Unicode characters in source files

### ISSUE THREAD

**Issue by Hugloss** [astral-sh/ruff#23934]
### Summary

# Feature Request: Configurable lint rule for unsafe Unicode characters in source files

**Category:** New lint rule
**Rule category:** `RUF`
**Suggested rule code:** `RUF300` *(or next available)*

## Problem

LLM-assisted edits and “smart” editors can silently insert Unicode characters that look harmless in many editors but break downstream tooling, especially codec-sensitive pipelines, `inspect-context`, and some Windows environments.

Examples we’ve seen in real projects include:

* `–` — en dash (`U+2013`)
* `—` — em dash (`U+2014`)
* `“` — left double quotation mark (`U+201C`)
* `”` — right double quotation mark (`U+201D`)
* `‘` — left single quotation mark (`U+2018`)
* `’` — right single quotation mark / apostrophe (`U+2019`)
* `-` — non-breaking hyphen (`U+2011`)
* `­` — soft hyphen (`U+00AD`)
* `\u009D` — control character (`U+009D`)
* `\ufeff` — UTF-8 BOM (`U+FEFF`)

The difficult part is that these characters are often visually subtle or invisible, so the source of the failure is not obvious.

## Proposal

Add a Ruff rule, tentatively **`RUF300`**, to detect suspicious Unicode characters in Python source files.

The goal is not to ban Unicode generally. The goal is to flag a narrow, configurable set of characters that are rarely intentional in source code and commonly introduced by copy/paste or editor auto-substitution.

### Default character set

A sensible default could be:

```python
DISALLOWED = {
    "\u2013": ("en dash", "Use ASCII hyphen-minus '-' instead"),
    "\u2014": ("em dash", "Use '--' or ASCII hyphen-minus '-' instead"),
    "\u201c": ("left double quotation mark", 'Use ASCII double quote `"` instead'),
    "\u201d": ("right double quotation mark", 'Use ASCII double quote `"` instead'),
    "\u2018": ("left single quotation mark", "Use ASCII apostrophe `'` instead"),
    "\u2019": ("right single quotation mark / apostrophe", "Use ASCII apostrophe `'` instead"),
    "\u2011": ("non-breaking hyphen", "Use ASCII hyphen-minus '-' instead"),
    "\u00ad": ("soft hyphen", "Remove or replace with ASCII hyphen '-'"),
    "\u009d": ("U+009D control character", "Remove this character"),
    "\ufeff": ("UTF-8 BOM", "Save the file as UTF-8 without BOM"),
}
```

## Example diagnostic

```text
src/mymodule/utils.py:42:17: RUF300 Suspicious Unicode character U+2013 (en dash).
  Use ASCII hyphen-minus '-' instead.
```

The diagnostic should include:

* file
* line
* column
* code point
* character description
* suggested replacement or remediation

## Configuration

The rule should be configurable so teams can extend or relax the default set.

Example:

```toml
[tool.ruff.lint.ruf300]
extend-disallowed = ["\u00a0"]  # non-breaking space
allow = ["\u2019"]             # permit right single quotation mark
```

If that naming does not fit Ruff’s existing config conventions, the same idea could be exposed through whatever structure is most consistent with existing rule-specific settings.

## Fixes

For characters with a clear ASCII replacement, Ruff could offer an unsafe fix under `--fix`.

Examples:

* `–` → `-`
* `—` → `--` or `-`
* `“` / `”` → `"`
* `‘` / `’` → `'`
* `-` → `-`

Characters like BOM or control characters may be report-only unless Ruff can safely rewrite them.

## Expected behavior

1. `ruff check` reports forbidden characters in Python source files.
2. Diagnostics are actionable and precise.
3. The rule is configurable.
4. `U+FEFF` is reported clearly when present in file content.
5. The rule applies to `.py` files, and ideally also `.pyi`.

## Non-goals

* This is not an “ASCII only” rule.
* This is not intended to ban legitimate Unicode in user-facing text.
* This is not a full replacement for byte-level encoding validation.

In particular, BOM-at-start detection may still require special handling if Ruff reads files after the codec layer has already normalized or stripped it.

## Acceptance criteria

* Forbidden Unicode characters are reported during `ruff check`.
* Diagnostics include file, line, column, code point, and replacement guidance.
* The rule is configurable.
* `U+FEFF` is handled clearly.
* Unsafe auto-fixes are available where replacements are obvious.
* These failures become easier to diagnose before they break downstream tools.

## Related work

* `RUF001`, `RUF002`, `RUF003`: Ruff already flags ambiguous Unicode in comments, docstrings, and strings. This proposal is similar in spirit, but targets a configurable set of tooling-hostile characters across source code more broadly.
* `pycodestyle` `W191`: precedent for low-level source character checks.
* PEP 672: related discussion around Unicode safety and source hygiene.

## Minimal reproduction

```python
def greet(name):
    """Say hello — the friendly way."""
    return f"Hello, {name}!"
```

A file containing characters like `—`, `“`, or `”` would produce diagnostics such as:

```text
bad.py:2:18: RUF300 Suspicious Unicode character U+2014 (em dash). Use '--' or ASCII hyphen-minus '-' instead.
```

**Comment by MichaReiser** [comment]
I'll merge this into https://github.com/astral-sh/ruff/issues/14433

I don't think a new rule makes sense only because an existing rule isn't configurable. We should make the existing rule configurable instead, assuming it's something we want.

**Comment by Hugloss** [comment]
That sounds right to me , I’m fine with folding this into the existing rule instead of adding a new one.

The real problem is agent loops.

A pretty normal setup now is:

* code agent writes code
* lint agent runs Ruff
* test agent runs tests
* check agent sends feedback back to code agent

When one of these Unicode landmines gets in, the loop stops doing useful work and starts burning turns on invisible text junk instead of code:

* smart quotes
* en/em dashes
* BOM
* odd control chars


That would save agent-to-agent workflows from getting stuck in dumb cleanup loops.
